Get saved statevectors

wagnersj opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [X] I have added the tests to cover my changes.
- [X] I have updated the documentation accordingly.
- [X] I have read the CONTRIBUTING document.
-->

### Summary

Changed `get_statevector()` in `qiskit.result.Result` to allow retrieval of a labelled statevector saved at a midpoint in a quantum circuit by `save_statevector()` in `qiskit.circuit.QuantumCircuit`.

Fixes #6706.


### Details and comments

A new optional argument, `label`, in `get_statevector()` allows users to get saved statevectors with the same label string used in calls to `save_statevector()`. This avoids the user from having to obtain the statevector from the raw data structure in a `Result` object.

jakelishman commented on qiskit/result/result.py:
```diff
@@ -305,27 +305,38 @@ def get_counts(self, experiment=None):
+        else:
+            _label = label
+
         try:
-            return postprocess.format_statevector(
-                self.data(experiment)["statevector"], decimals=decimals
-            )
+            return postprocess.format_statevector(self.data(experiment)[_label], decimals=decimals)
         except KeyError as ex:
-            raise QiskitError(f'No statevector for experiment "{repr(experiment)}"') from ex
+            if label is None:
+                raise QiskitError(f'No statevector for experiment "{repr(experiment)}"') from ex
+            raise QiskitError(
+                f'No statevector "{label}"' 'for experiment "{repr(experiment)}"'
+            ) from ex
```
In line 338, the string has been split in two, and consequently the f-string modifier isn't being applied to the second half, which also needs it.

Also, while an incredibly nitpicky point, I first thought that the two errors being raised were different because of the branching.  Would it perhaps be a shade clearer that they're the same if we do something like
```python
label_text = "" if label is None else (label + " ")
raise QiskitError(
    f'No statevector {label_text}for experiment "{experiment!r}"'
) from ex
```
(i.e. branch on the message, rather than the raise)

jakelishman reviewed (changes requested):
Thanks for doing this - it looks like a sensible change to me.  I left a comment about a bug in the error message, but other than that this looks neat.

jakelishman reviewed (approved):
Looks good to me.  I just pushed up a re-instatement of the quote marks around the label (my bad - I had omitted them in my suggestion).  Since it would be the first time I'd merge a new feature, I'll let someone else check it over as well to make sure it's sensible and nothing's missing, but seems useful to me with no real downsides.

wagnersj commented:
@jakelishman I completely agree with your comments. I'll get to work on your requests.

wagnersj commented:
I have added the same functionality to the `get_unitary` method to retrieve a saved unitary with a label. 

For checking the data, I added some code in the `format_statevector` and `format_unitary` methods to check if the the given data has the correct shape. I also set the datatypes of the arguments to these methods so that it will throw an exception if the data is not a supported datatype.

docstrings have also been updated.

jakelishman commented:
Sorry again for the delay.  @chriseclectic is thinking about changing over how some of this stuff is handled on the Aer side, so I'll let him look at this, but in principle from the Terra side it seems ok.

aburousan commented:
> ### Summary
> Changed `get_statevector()` in `qiskit.result.Result` to allow retrieval of a labelled statevector saved at a midpoint in a quantum circuit by `save_statevector()` in `qiskit.circuit.QuantumCircuit`.
> 
> Fixes #6706.
> 
> ### Details and comments
> A new optional argument, `label`, in `get_statevector()` allows users to get saved statevectors with the same label string used in calls to `save_statevector()`. This avoids the user from having to obtain the statevector from the raw data structure in a `Result` object.
![Screenshot from 2022-07-01 00-17-36](https://user-images.githubusercontent.com/60939745/176755141-336cfdac-321c-49c5-894c-45c3893a88b0.png)

This is not working can you tell why?

wagnersj commented:
@aburousan, note that this pull request has not been merged into the `main` branch of Qiskit Terra yet. Are you using a released version of Qiskit, or are you using the `wagnersj:get_statevector` branch from source?

aburousan commented:
> 

No..no...I thought that it is merged... Alright
